lib/wallaroo/core/sink/connector_sink/connector_sink.pony

```diff
@@ -363,49 +481,197 @@ actor ConnectorSink is Sink
     end
     match barrier_token
     | let sbt: CheckpointBarrierToken =>
+      if not _notify.twopc_intro_done then
+        // This sink applies Pony runtime backpressure when disconnected,
+        // so it's quite unlikely that we will get here: sending
+        // messages to us will block the senders.  However, the nature
+        // of backpressure may change over time as the runtime's
+        // backpressure system changes.
+        ifdef "checkpoint_trace" then
+          @printf[I32]("2PC: preemptive abort: connector sink not fully connected\n".cstring())
+        end
+        abort_decision("connector sink not fully connected",
+          _twopc.txn_id, _twopc.barrier_token)
+        _twopc.preemptive_txn_abort(sbt)
+        return
+      end
+
       checkpoint_state(sbt.id)
+      _notify.twopc_txn_id_current = _twopc.txn_id
+      match _twopc.barrier_complete(sbt)
+      | None =>
+        // If no data has been processed by the sink since the last
+        // checkpoint, then don't bother with 2PC, return early.
+
+        _barrier_coordinator.ack_barrier(this, sbt)
+        ifdef "checkpoint_trace" then
+          @printf[I32]("2PC: no data written during this checkpoint interval, skipping 2PC round\n".cstring())
+        end
+        return
+
+      | let msg: cp.MessageMsg =>
+        _notify.send_msg(this, msg)
+        ifdef "checkpoint_trace" then
+          @printf[I32]("2PC: sent phase 1 for txn_id %s\n".cstring(), _twopc.txn_id.cstring())
+        end
+      end
+
+      // As a 2PC participant as a Wallaroo *sink*, we cannot
```

Comment:
This comment is ok as far as the types of external sinks you've been designing for are concerned, but I think it's misleading here.  That's because this does not apply to external connectors that can deal with speculative messages.  For example, if the external connector was writing to a file, precommit could mean writing the outputs to a file with the checkpoint number in the precommit filename.  It could begin writing the precommit file for the next checkpoint as soon as those outputs are available.  On abort of a checkpoint, it could delete the precommit files for that checkpoint and all subsequent checkpoints.

I think the comment should reflect the fact that for now we've chosen to optimize for the case where the external connector cannot do anything with speculative outputs other than queue them.